### PR TITLE
DMA Endpoint Mix-ins and some package reshuffling

### PR DESCRIPTION
**Pull request by davidbiancolin**
This PR does three things, to pave the way for print & trace synthesis. 

1) Makes Widget extend MultiIOModule (yay!)
2) It takes the DMA-side hardware (BRAM-hosted queues) and moves them into mixins that be added to a Widget
3) Reorganizes the file structure in midas.widget into something more sensible
   - Endpoint and EndpointWidget are put in the `widget/Endpoints.scala` from `core/Endpoints.scala`
   - `widget/NastiIO.scala` only contains AXI-4 related endpoint code
   - DMA related code is moved into `widget/DMATransport.scala`

@zhemao @sagark RFC

### CODE REVIEW

File: src/main/scala/midas/widgets/DMATransport.scala [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  } .otherwise {
+    outgoingCount := outgoingCount
+  }
+
+  // check to see if pcis has valid output instead of waiting for timeouts
+  attach(outgoingPCISdat.io.deq.valid, "pcis_out_valid", ReadOnly)
+  attach(outgoingCount, "outgoing_count", ReadOnly)
+
+  // TODO, will these queues bottleneck us? - Biancolin: no. -> just read out of BRAM queue?
+  val ar_queue = Queue(dma.ar, 10)
```
Comment (davidbiancolin) [inline src/main/scala/midas/widgets/DMATransport.scala]: @sagark What's the rationale for these? They are pretty enormous, and i think additional DMA buffering (for timing closure) can be moved into FPGATop (they should also be PipeQueues i think).

**Comment by davidbiancolin** [comment]
This is ready for review @sagark, @zhemao

File: src/main/scala/midas/passes/Utils.scala [hunk 2]
```diff
@@ -75,6 +94,19 @@ object Utils {
...
+  def inputForm = HighForm
+  def outputForm = HighForm
+  override def name = s"[MIDAS] Debugging Emission Pass: $fileName"
+
+  def execute(state: CircuitState) = {
+    Utils.writeState(state, fileName)
+    state
+  }
+}
+
```
Comment (zhemao) [inline src/main/scala/midas/passes/Utils.scala]: What exactly are the changes in this file for?

File: src/main/scala/midas/widgets/Print.scala [hunk 3]
```diff
@@ -55,102 +54,48 @@ class PrintRecordEndpoint extends Endpoint {
...
+  //  sb.append(genMacro("PRINTS_COUNT_ADDR", UInt32(base + countAddrs.head)))
+  //  sb.append(genArray("PRINTS_WIDTHS", printPort.elements.toSeq map (x => UInt32(x._2.getWidth))))
+  //  if (!p(HasDMAChannel)) {
+  //    sb.append(genArray("PRINTS_DATA_ADDRS", printAddrs.toSeq map (x => UInt32(base + x))))
+  //  } else {
+  //    sb.append(genMacro("HAS_DMA_CHANNEL"))
+  //  }
+  //}
+
+  //genCRFile()
```
Comment (zhemao) [inline src/main/scala/midas/widgets/Print.scala]: Why is this all commented out now?

Comment (davidbiancolin) [reply src/main/scala/midas/widgets/Print.scala]: Print synthesis is not functional yet, so it was easier to comment out the code for now.

File: src/main/scala/midas/widgets/DMATransport.scala (hunk 1 above)
Comment (sagark) [reply src/main/scala/midas/widgets/DMATransport.scala]: Yeah, those are leftover from the initial NIC widget implementation that didn't use DMA. They can be removed.

Comment (davidbiancolin) [reply src/main/scala/midas/widgets/DMATransport.scala]: I'll remove them as part of PrintSynthesis bring up.

**Pull request merged**